### PR TITLE
Loki Query Editor: Add support to display query parsing errors to users

### LINKED ISSUE #57746: Loki: Query validation in monaco editor
### Summary

Our users would benefit greatly if they instantly learned if the query they constructed is valid or not. Moreover if they knew where exactly does query have error, it would be significantly easier and faster to fix it. This would result in users who are more confident in queries that they are sending to Loki. 

### Problem statement

Currently, to learn if the created LogQL is valid, users need to run the query and if the query is invalid, they get the not-super-user-friendly error message. With a new monaco editor we have a great opportunity to improve query writing experience for our users by doing query validation and instantly showing query errors in editor. This way, users learn right away which part of the query is incorrect. This removes frustration of running a query, waiting for result and getting error message about invalid logql. 

### Scope

Design Doc

### Solution

With loki lezer parser and monaco editor, we have a great way how to improve query writing. While now, users have to send query to see if it is valid or invalid, we could use parser to show them this information even before sending the query out. Moreover, we could point them to the exact spot where the issue it.

This requires design doc as this could be a new data source API used for all data sources using monaco editor. 

Image

Image

### Measurement

- Higher ratio of successful queries.
- Less queries that error out due to incorrect LogQl. 

### Resources

https://github.com/grafana/grafana/discussions/52358
https://github.com/grafana/grafana/discussions/53647

### PR DESCRIPTION
**Pull request by matyax**
**What is this feature?**

This PR introduces a new feature to the Loki Monaco Query Editor to validate queries before sending them to Loki. To achieve that, we use Lezer to parse the query and, if there are error nodes, we process them to be displayed as errors in the editor.

It supports any Loki query, even single o multi-line.

**Which issue(s) does this PR fix?**:

Fixes https://github.com/grafana/grafana/issues/57746

**Special notes for your reviewer**:

Demo:

https://user-images.githubusercontent.com/1069378/204602720-04fb6cbf-c560-4013-9d3f-7095c94ceff8.mov

Try different combinations of incorrect queries and see the editor should point to incorrect parts of it.

The error should match the error responses from Loki:

imagen

imagen

imagen

imagen

### CODE REVIEW

**Review by ivanahuckova — COMMENTED** [review]
This looks very good! I have found 1 issue though - currently, this marks internal grafana variables as parsing errors. In validate, we should consider the fact, that users can use `grafana variables` and `template variables` and those are not errors.  

image

**Comment by matyax** [comment]
> we should consider the fact, that users can use grafana variables and template variables and those are not errors

@ivanahuckova I'm not really sure on how to approach that. Do you have any suggestion?

**Comment by matyax** [comment]
After reviewing this, I found no straightforward way to approach this.

## Interpolate and parse

We can interpolate user variables and bypass the $token error, but the disadvantage is that the error reporting becomes less accurate in situations where, for example, the string length of the variable and the content are really different. E.g.:

imagen

In this case, the variable name is longer than the content, so it partially highlights the error.

imagen

imagen

imagen

The advantage of this approach is that the query parsing uses the interpolated query, so we can give feedback to the user when the query with variables is incorrect, for example `rate({place="luna"}[$__from])`

imagen

Also, there's no straighforward way of handling `$__interval` because it's excluded from the interpolation, so we would have to either ignore it in `rate({place="luna"}[$__interval])` or force its interpolation with a default value (e.g. `1m`), which could be potentially confusing for the user.

## Ignore variables

The alternative approach would be to ignore errors that could be variables. The disadvantages are having a less accurate parsing of the query, and there's no straightforward way to identify  variables, so we would have to guess or add enough complexity to detect actual variable names in errors.

**Comment by ivanahuckova** [comment]
I have couple ideas that might be worth looking into:

**In Explore** the issues are basically only internal built in Grafana variables (such as `$__interval`, `$__range`, `$__interval_ms`). So in case of an error in `Duration` (but we should double check if there are any other places in query where built in variables could be used), we could check if it is internal built in variable and then ignore error.  Here is the list of built in variables that could be considered:  https://grafana.com/docs/grafana/v9.0/variables/variable-types/global-variables/ .

**In Dashboard** we have the same issue plus issue with template variables (e.g. $label="$value"). I was brainstorming some ideas and the best I think I had is using `lokiDatasource.getVariables()` function that returns list of template variable names. And we could just check if the error node is template variable and then just ignore the errors. I wouldn't at this point be looking into the value, and just consider using any template variable as "correct"

Another ideas I had/I was looking into:

1. We could consider interpolating variables. I did a quick test and we could do something like (but this is going to mess up highlight because of the miss-match between length of variable and actual value 😞 ):

```
  const interpolatedQuery = datasource.interpolateVariablesInQueries([{ expr: query, refId: '' }], {});
  const tree = parser.parse(interpolatedQuery[0].expr);
  tree.iterate({
    enter: (node): false | void => {
      if (node.type.id === ErrorId) {
        errorNodes.push(node.node);
      }
    },
  });
```

2. Maybe we could do something similar as we do in builder where we check for variables (based on regex) and replace them. https://github.com/grafana/grafana/blob/main/public/app/plugins/datasource/prometheus/querybuilder/shared/parsingUtils.ts#L38.

**Comment by matyax** [comment]
@ivanahuckova Made an update. Wanna take a look at let me know how it feels now? @grafana/observability-logs feel free to try it as well if you have some time to spare.

File: public/app/plugins/datasource/loki/components/monaco-query-field/MonacoQueryField.tsx [hunk 1]
```diff
@@ -125,8 +130,13 @@ const MonacoQueryField = ({ languageProvider, history, onBlur, onRunQuery, initi
...
             }
-
-            const errors = validateQuery(model.getValue(), model.getLinesContent()) || [];
+            const query = model.getValue();
+            const errors =
+              validateQuery(
+                query,
+                datasource.interpolateString(query, placeHolderScopedVars),
+                model.getLinesContent()
+              ) || [];
```
Comment (ivanahuckova) [inline public/app/plugins/datasource/loki/components/monaco-query-field/MonacoQueryField.tsx:139]: The issue with this approach is that if we interpolate values and they have different length than their variable, the error highlight is at wrong position 
image
image

So we should probably adjust the position, if you want to validate also variables.

Comment (ivanahuckova) [reply public/app/plugins/datasource/loki/components/monaco-query-field/MonacoQueryField.tsx:139]: But I personally think, that we could just "ignore" template variables and maybe decide in the future if we want to enhance validation to include interpolating of variables.

Comment (matyax) [reply public/app/plugins/datasource/loki/components/monaco-query-field/MonacoQueryField.tsx:139]: Thanks for the feedback. I have a few more ideas to explore, so I'll give this another try. As a last resource, we can ignore variables and consider variables as a future enhancement.

Comment (matyax) [reply public/app/plugins/datasource/loki/components/monaco-query-field/MonacoQueryField.tsx:139]: @ivanahuckova I changed my approach a little bit, and got more accurate results when validating interpolated queries.

I still think we should support it, to provide the best experience validating queries (i.e. minimizing the amount of failed queries reaching the backend), so I look forward to keep iterating here.

**Review by ivanahuckova — APPROVED** [review]
LGTM, this works great! Just`$__range` needs to be added to scopedVars, as it is marked as error, even though it is offered as suggestion in autocomplete.

**Comment by matyax** [comment]
Done, good catch!

**Pull request merged**